File: javascript/lib/native.js
// Centralized koffi binding to the native Selenium core library.
//
// 1:1 with the `aether_sel_embed_*` C ABI exported by the in-repo
// `core/embed.ae` (the engine itself is `core/selenium_core.ae`). The opaque
// session handle is a `void*`; NULL means open failed. Every `char*` returned
// by the ABI is caller-owned and NUL-terminated — `takeString` decodes it into
// a JS string and then frees it via `aether_sel_embed_free_string`, per the
// ABI's ownership rule.
//
// Handle-based contract (matching the Aether side): N independent sessions can
// run concurrently in one process, each keyed by its own handle; every execute
// / accessor call takes that handle.

'use strict'

const fs = require('fs')
const path = require('path')
const koffi = require('koffi')

const LIB_BASE = 'selenium_core'

function fileName() {
  switch (process.platform) {
    case 'win32':
      return `${LIB_BASE}.dll`
    case 'darwin':
      return `lib${LIB_BASE}.dylib`
    default:
      return `lib${LIB_BASE}.so`
  }
}

let explicitPath = null
let loadedLib = null

/**
 * Pin an explicit path to the native library, used at first load. Wins over the
 * bundled `native/` default and the SELENIUM_CORE_LIB env override. No-op once
 * loaded.
 */
function configure(nativeLib) {
  if (nativeLib && !loadedLib) explicitPath = nativeLib
}

function resolveLibraryPath() {
  if (explicitPath && fs.existsSync(explicitPath)) return explicitPath
  const override = process.env.SELENIUM_CORE_LIB
  if (override && fs.existsSync(override)) return override
  const bundled = path.resolve(__dirname, '..', 'native', fileName())
  if (fs.existsSync(bundled)) return bundled
  return fileName()
}

function lib() {
  if (!loadedLib) loadedLib = koffi.load(resolveLibraryPath())
  return loadedLib
}

// Lazily bind a native function so a pinned explicit path takes effect.
function lazy(signature) {
  let f = null
  return (...args) => {
    if (!f) f = lib().func(signature)
    return f(...args)
  }
}

// ---- lifecycle ----
const open = lazy('void* aether_sel_embed_open(const char* base_url)')
const close = lazy('void aether_sel_embed_close(void* h)')

// ---- workhorse ----
const execute = lazy('int aether_sel_embed_execute(void* h, const char* name, const char* params_json)')

// ---- result accessors ----
const lastValue = lazy('void* aether_sel_embed_last_value(void* h)')
const lastStatus = lazy('int aether_sel_embed_last_status(void* h)')
const lastErrorCode = lazy('int aether_sel_embed_last_error_code(void* h)')
const lastError = lazy('void* aether_sel_embed_last_error(void* h)')
const sessionId = lazy('void* aether_sel_embed_session_id(void* h)')

// ---- pure helpers ----
const byLocator = lazy('void* aether_sel_embed_by_locator(const char* strategy, const char* value)')
const route = lazy('void* aether_sel_embed_route(const char* name)')
const buildRequest = lazy('void* aether_sel_embed_build_request(const char* name, const char* session_id, const char* params_json)')
const errorCode = lazy('int aether_sel_embed_error_code(const char* w3c_error)')

// ---- atom-backed commands (a shared JS atom run in-page by the engine) ----
// isDisplayed / getAttribute / relative-locator all execute the same JS atoms
// upstream Selenium ships, via the engine's executeScript path. Results drain
// through the usual last_value channel. char* args/returns go through void*.
const executeAtom = lazy(
  'int aether_sel_embed_execute_atom(void* h, const char* atom, const char* elem_id, const char* extra_json)',
)
const isDisplayed = lazy('int aether_sel_embed_is_displayed(void* h, const char* elem_id)')
const getAttribute = lazy('int aether_sel_embed_get_attribute(void* h, const char* elem_id, const char* name)')
const atomStrArg = lazy('void* aether_sel_embed_atom_str_arg(const char* s)')
const findRelative = lazy('int aether_sel_embed_find_relative(void* h, const char* base_css, const char* filters_json)')

// ---- TLS config (per session handle; set before newSession) ----
const setCa = lazy('void aether_sel_embed_set_ca(void* h, const char* ca_path)')
const setInsecure = lazy('void aether_sel_embed_set_insecure(void* h, int on)')

// ---- driver orchestration (spawn/adopt a driver process in-binding) ----
// An opaque driver handle, independent of the W3C session handle. launch_driver
// and ensure_driver return that handle (void*); driver_url returns a caller-owned
// char* (as void*, taken via takeString); driver_pid returns int.
const resolveDriver = lazy('void* aether_sel_embed_resolve_driver(const char* browser, const char* hint)')
const launchDriver = lazy('void* aether_sel_embed_launch_driver(const char* driver_path, int timeout_ms)')
const ensureDriver = lazy('void* aether_sel_embed_ensure_driver(const char* browser, const char* hint, int timeout_ms)')
const driverUrl = lazy('void* aether_sel_embed_driver_url(void* dh)')
const driverPid = lazy('int aether_sel_embed_driver_pid(void* dh)')
const stopDriver = lazy('void aether_sel_embed_stop_driver(void* dh)')

// ---- WebDriver-BiDi (over the session's webSocketUrl) ----
// An opaque BiDi channel handle, independent of the W3C session handle.
const bidiOpen = lazy('void* aether_sel_embed_bidi_open(const char* ws_url)')
const bidiClose = lazy('void aether_sel_embed_bidi_close(void* h)')
const bidiSend = lazy('int aether_sel_embed_bidi_send(void* h, int id, const char* method, const char* params_json)')
const bidiPump = lazy('int aether_sel_embed_bidi_pump(void* h, int timeout_ms)')
const bidiFd = lazy('int aether_sel_embed_bidi_fd(void* h)')
const bidiPollReply = lazy('void* aether_sel_embed_bidi_poll_reply(void* h, int id)')
const bidiPollEvent = lazy('void* aether_sel_embed_bidi_poll_event(void* h)')
const bidiLostEvents = lazy('int aether_sel_embed_bidi_lost_events(void* h)')
const bidiCancel = lazy('void aether_sel_embed_bidi_cancel(void* h, int id)')
const bidiSubscribe = lazy('void* aether_sel_embed_bidi_subscribe(void* h, int id, const char* events_csv, int timeout_ms)')
const bidiUnsubscribe = lazy('void* aether_sel_embed_bidi_unsubscribe(void* h, int id, const char* events_csv, int timeout_ms)')
const bidiWaitEvent = lazy('void* aether_sel_embed_bidi_wait_event(void* h, const char* method, int timeout_ms)')

// ---- typed BiDi convenience commands (char* replies returned as void*) ----
const bidiGetTree = lazy('void* aether_sel_embed_bidi_get_tree(void* h, int id, int timeout_ms)')
const bidiScriptEvaluate = lazy(
  'void* aether_sel_embed_bidi_script_evaluate(void* h, int id, const char* expr, const char* context_id, int timeout_ms)',
)
const bidiNavigate = lazy(
  'void* aether_sel_embed_bidi_navigate(void* h, int id, const char* context_id, const char* url, int timeout_ms)',
)

// ---- BiDi network interception (char* replies returned as void*) ----
const bidiNetworkAddIntercept = lazy(
  'void* aether_sel_embed_bidi_network_add_intercept(void* h, int id, const char* phases_csv, const char* url_pattern, int timeout_ms)',
)
const bidiNetworkRemoveIntercept = lazy(
  'void* aether_sel_embed_bidi_network_remove_intercept(void* h, int id, const char* intercept_id, int timeout_ms)',
)
const bidiNetworkContinueRequest = lazy(
  'void* aether_sel_embed_bidi_network_continue_request(void* h, int id, const char* request_id, int timeout_ms)',
)
const bidiNetworkFailRequest = lazy(
  'void* aether_sel_embed_bidi_network_fail_request(void* h, int id, const char* request_id, int timeout_ms)',
)
const bidiNetworkProvideResponse = lazy(
  'void* aether_sel_embed_bidi_network_provide_response(void* h, int id, const char* request_id, int status, const char* content_type, const char* body, int timeout_ms)',
)
const bidiNetworkContinueWithAuth = lazy(
  'void* aether_sel_embed_bidi_network_continue_with_auth(void* h, int id, const char* request_id, const char* username, const char* password, int timeout_ms)',
)
const bidiNetworkSetCacheBehavior = lazy(
  'void* aether_sel_embed_bidi_network_set_cache_behavior(void* h, int id, const char* behavior, int timeout_ms)',
)

// ---- string ownership ----
const freeString = lazy('void aether_sel_embed_free_string(void* s)')

function isNull(ptr) {
  return !ptr || koffi.address(ptr) === 0n
}

// Decode a caller-owned native char* into a JS string, then free it. "" on NULL.
function takeString(ptr) {
  if (isNull(ptr)) return ''
  try {
    return koffi.decode(ptr, 'char', -1)
  } finally {
    freeString(ptr)
  }
}

module.exports = {
  configure,
  open,
  close,
  execute,
  lastValue,
  lastStatus,
  lastErrorCode,
  lastError,
  sessionId,
  executeAtom,
  isDisplayed,
  getAttribute,
  atomStrArg,
  findRelative,
  byLocator,
  route,
  buildRequest,
  errorCode,
  setCa,
  setInsecure,
  resolveDriver,
  launchDriver,
  ensureDriver,
  driverUrl,
  driverPid,
  stopDriver,
  bidiOpen,
  bidiClose,
  bidiSend,
  bidiPump,
  bidiFd,
  bidiPollReply,
  bidiPollEvent,
  bidiLostEvents,
  bidiCancel,
  bidiSubscribe,
  bidiUnsubscribe,
  bidiWaitEvent,
  bidiGetTree,
  bidiScriptEvaluate,
  bidiNavigate,
  bidiNetworkAddIntercept,
  bidiNetworkRemoveIntercept,
  bidiNetworkContinueRequest,
  bidiNetworkFailRequest,
  bidiNetworkProvideResponse,
  bidiNetworkContinueWithAuth,
  bidiNetworkSetCacheBehavior,
  takeString,
  isNull,
}

File: dart/lib/src/native.dart
/// The 1:1 dart:ffi symbol table for the Selenium core C ABI (`core/embed.ae`,
/// built on the pure-Aether `core/selenium_core.ae` engine). This library is the
/// ONLY place in the Dart binding that knows about the C ABI; everything above
/// it is idiomatic Dart. No protocol logic lives here or anywhere in this
/// package — the engine is shared by every language binding.
///
/// `core/embed.ae` names its exports `sel_embed_<name>`; `--emit=lib` mangles
/// them to `aether_sel_embed_<name>`, the names looked up here. Every `char*`
/// this ABI returns is caller-owned and must be freed via
/// `aether_sel_embed_free_string` — [takeString] does that.
library;

import 'dart:ffi' as ffi;
import 'dart:io' show Platform;

import 'package:ffi/ffi.dart' as pkgffi;

typedef _OpenC = ffi.Pointer<ffi.Void> Function(ffi.Pointer<pkgffi.Utf8>);
typedef _Open = ffi.Pointer<ffi.Void> Function(ffi.Pointer<pkgffi.Utf8>);

typedef _CloseC = ffi.Void Function(ffi.Pointer<ffi.Void>);
typedef _Close = void Function(ffi.Pointer<ffi.Void>);

typedef _ExecuteC = ffi.Int Function(
    ffi.Pointer<ffi.Void>, ffi.Pointer<pkgffi.Utf8>, ffi.Pointer<pkgffi.Utf8>);
typedef _Execute = int Function(
    ffi.Pointer<ffi.Void>, ffi.Pointer<pkgffi.Utf8>, ffi.Pointer<pkgffi.Utf8>);

typedef _HandleToStrC = ffi.Pointer<pkgffi.Utf8> Function(ffi.Pointer<ffi.Void>);
typedef _HandleToStr = ffi.Pointer<pkgffi.Utf8> Function(ffi.Pointer<ffi.Void>);

typedef _HandleToIntC = ffi.Int Function(ffi.Pointer<ffi.Void>);
typedef _HandleToInt = int Function(ffi.Pointer<ffi.Void>);

typedef _StrToStrC = ffi.Pointer<pkgffi.Utf8> Function(ffi.Pointer<pkgffi.Utf8>);
typedef _StrToStr = ffi.Pointer<pkgffi.Utf8> Function(ffi.Pointer<pkgffi.Utf8>);

typedef _Str2ToStrC = ffi.Pointer<pkgffi.Utf8> Function(
    ffi.Pointer<pkgffi.Utf8>, ffi.Pointer<pkgffi.Utf8>);
typedef _Str2ToStr = ffi.Pointer<pkgffi.Utf8> Function(
    ffi.Pointer<pkgffi.Utf8>, ffi.Pointer<pkgffi.Utf8>);

typedef _StrToIntC = ffi.Int Function(ffi.Pointer<pkgffi.Utf8>);
typedef _StrToInt = int Function(ffi.Pointer<pkgffi.Utf8>);

typedef _FreeStringC = ffi.Void Function(ffi.Pointer<pkgffi.Utf8>);
typedef _FreeString = void Function(ffi.Pointer<pkgffi.Utf8>);

// ---- WebDriver-BiDi (over the session's webSocketUrl) ----
// The BiDi channel handle (void*) is opaque and independent of the W3C handle.
typedef _BidiPumpC = ffi.Int Function(ffi.Pointer<ffi.Void>, ffi.Int32);
typedef _BidiPump = int Function(ffi.Pointer<ffi.Void>, int);

typedef _BidiSendC = ffi.Int Function(ffi.Pointer<ffi.Void>, ffi.Int32,
    ffi.Pointer<pkgffi.Utf8>, ffi.Pointer<pkgffi.Utf8>);
typedef _BidiSend = int Function(
    ffi.Pointer<ffi.Void>, int, ffi.Pointer<pkgffi.Utf8>, ffi.Pointer<pkgffi.Utf8>);

typedef _BidiPollReplyC = ffi.Pointer<pkgffi.Utf8> Function(
    ffi.Pointer<ffi.Void>, ffi.Int32);
typedef _BidiPollReply = ffi.Pointer<pkgffi.Utf8> Function(
    ffi.Pointer<ffi.Void>, int);

typedef _BidiCancelC = ffi.Void Function(ffi.Pointer<ffi.Void>, ffi.Int32);
typedef _BidiCancel = void Function(ffi.Pointer<ffi.Void>, int);

typedef _BidiSubC = ffi.Pointer<pkgffi.Utf8> Function(
    ffi.Pointer<ffi.Void>, ffi.Int32, ffi.Pointer<pkgffi.Utf8>, ffi.Int32);
typedef _BidiSub = ffi.Pointer<pkgffi.Utf8> Function(
    ffi.Pointer<ffi.Void>, int, ffi.Pointer<pkgffi.Utf8>, int);

typedef _BidiWaitEventC = ffi.Pointer<pkgffi.Utf8> Function(
    ffi.Pointer<ffi.Void>, ffi.Pointer<pkgffi.Utf8>, ffi.Int32);
typedef _BidiWaitEvent = ffi.Pointer<pkgffi.Utf8> Function(
    ffi.Pointer<ffi.Void>, ffi.Pointer<pkgffi.Utf8>, int);

/// A loaded engine: the [ffi.DynamicLibrary] plus every symbol bound once.
class Native {
  final _Open open;
  final _Close close;
  final _Execute execute;
  final _HandleToStr lastValue;
  final _HandleToInt lastStatus;
  final _HandleToInt lastErrorCode;
  final _HandleToStr lastError;
  final _HandleToStr sessionId;
  final _Str2ToStr byLocator;
  final _StrToStr route;
  final _StrToInt errorCode;
  final _FreeString freeString;

  // ---- WebDriver-BiDi ----
  final _Open bidiOpen;
  final _Close bidiClose;
  final _BidiSend bidiSend;
  final _BidiPump bidiPump;
  final _HandleToInt bidiFd;
  final _BidiPollReply bidiPollReply;
  final _HandleToStr bidiPollEvent;
  final _HandleToInt bidiLostEvents;
  final _BidiCancel bidiCancel;
  final _BidiSub bidiSubscribe;
  final _BidiSub bidiUnsubscribe;
  final _BidiWaitEvent bidiWaitEvent;

  Native._(ffi.DynamicLibrary lib)
      : open = lib.lookupFunction<_OpenC, _Open>('aether_sel_embed_open'),
        close = lib.lookupFunction<_CloseC, _Close>('aether_sel_embed_close'),
        execute =
            lib.lookupFunction<_ExecuteC, _Execute>('aether_sel_embed_execute'),
        lastValue = lib.lookupFunction<_HandleToStrC, _HandleToStr>(
            'aether_sel_embed_last_value'),
        lastStatus = lib.lookupFunction<_HandleToIntC, _HandleToInt>(
            'aether_sel_embed_last_status'),
        lastErrorCode = lib.lookupFunction<_HandleToIntC, _HandleToInt>(
            'aether_sel_embed_last_error_code'),
        lastError = lib.lookupFunction<_HandleToStrC, _HandleToStr>(
            'aether_sel_embed_last_error'),
        sessionId = lib.lookupFunction<_HandleToStrC, _HandleToStr>(
            'aether_sel_embed_session_id'),
        byLocator = lib.lookupFunction<_Str2ToStrC, _Str2ToStr>(
            'aether_sel_embed_by_locator'),
        route =
            lib.lookupFunction<_StrToStrC, _StrToStr>('aether_sel_embed_route'),
        errorCode = lib.lookupFunction<_StrToIntC, _StrToInt>(
            'aether_sel_embed_error_code'),
        freeString = lib.lookupFunction<_FreeStringC, _FreeString>(
            'aether_sel_embed_free_string'),
        bidiOpen =
            lib.lookupFunction<_OpenC, _Open>('aether_sel_embed_bidi_open'),
        bidiClose =
            lib.lookupFunction<_CloseC, _Close>('aether_sel_embed_bidi_close'),
        bidiSend = lib.lookupFunction<_BidiSendC, _BidiSend>(
            'aether_sel_embed_bidi_send'),
        bidiPump = lib.lookupFunction<_BidiPumpC, _BidiPump>(
            'aether_sel_embed_bidi_pump'),
        bidiFd = lib.lookupFunction<_HandleToIntC, _HandleToInt>(
            'aether_sel_embed_bidi_fd'),
        bidiPollReply = lib.lookupFunction<_BidiPollReplyC, _BidiPollReply>(
            'aether_sel_embed_bidi_poll_reply'),
        bidiPollEvent = lib.lookupFunction<_HandleToStrC, _HandleToStr>(
            'aether_sel_embed_bidi_poll_event'),
        bidiLostEvents = lib.lookupFunction<_HandleToIntC, _HandleToInt>(
            'aether_sel_embed_bidi_lost_events'),
        bidiCancel = lib.lookupFunction<_BidiCancelC, _BidiCancel>(
            'aether_sel_embed_bidi_cancel'),
        bidiSubscribe = lib.lookupFunction<_BidiSubC, _BidiSub>(
            'aether_sel_embed_bidi_subscribe'),
        bidiUnsubscribe = lib.lookupFunction<_BidiSubC, _BidiSub>(
            'aether_sel_embed_bidi_unsubscribe'),
        bidiWaitEvent = lib.lookupFunction<_BidiWaitEventC, _BidiWaitEvent>(
            'aether_sel_embed_bidi_wait_event');

  static Native? _instance;
  static String? _explicitPath;

  /// Pin an explicit library path (wins over env/bundled).
  static void configure(String? path) {
    if (path != null && path.isNotEmpty && _instance == null) {
      _explicitPath = path;
    }
  }

  static Native get instance => _instance ??= Native._(_load());

  static String _fileName() {
    if (Platform.isWindows) return 'selenium_core.dll';
    if (Platform.isMacOS) return 'libselenium_core.dylib';
    return 'libselenium_core.so';
  }

  static ffi.DynamicLibrary _load() {
    for (final candidate in _candidates()) {
      try {
        return ffi.DynamicLibrary.open(candidate);
      } catch (_) {
        // try next
      }
    }
    // Bare name (OS loader).
    return ffi.DynamicLibrary.open(_fileName());
  }

  static Iterable<String> _candidates() sync* {
    if (_explicitPath != null && _explicitPath!.isNotEmpty) yield _explicitPath!;
    final env = Platform.environment['SELENIUM_CORE_LIB'];
    if (env != null && env.isNotEmpty) yield env;
    // Bundled next to the package: <pkg>/native/<lib>. Resolve relative to the
    // script's package when possible; else a plain relative path.
    yield 'native/${_fileName()}';
  }

  /// Copy a caller-owned native char* into a Dart String, then free it.
  String takeString(ffi.Pointer<pkgffi.Utf8> ptr) {
    if (ptr == ffi.nullptr) return '';
    try {
      return ptr.toDartString();
    } finally {
      freeString(ptr);
    }
  }
}
